File: chazm-api/src/main/java/runtimemodels/chazm/api/relation/manager/AchievesManager.java
package runtimemodels.chazm.api.relation.manager;

import runtimemodels.chazm.api.entity.Role;
import runtimemodels.chazm.api.entity.SpecificationGoal;
import runtimemodels.chazm.api.id.RoleId;
import runtimemodels.chazm.api.id.SpecificationGoalId;
import runtimemodels.chazm.api.id.UniqueId;

import java.util.Set;

/**
 * The {@linkplain AchievesManager} interface defines the APIs for managing achieves relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface AchievesManager {
    /**
     * Creates an achieves relation between a {@linkplain Role} and a {@linkplain SpecificationGoal} in this {@linkplain AchievesManager}.
     *
     * @param roleId the {@linkplain UniqueId} identifying the {@linkplain Role} for the achieves relation.
     * @param goalId the {@linkplain UniqueId} identifying the {@linkplain SpecificationGoal} for the achieves relation.
     */
    void addAchieves(RoleId roleId, SpecificationGoalId goalId);

    /**
     * Returns a set of {@linkplain SpecificationGoal}s that is achieved by a {@linkplain Role} from this {@linkplain AchievesManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @return a set of {@linkplain SpecificationGoal}s.
     */
    Set<SpecificationGoal> getAchieves(RoleId id);

    /**
     * Returns a set of {@linkplain Role}s that achieves a {@linkplain SpecificationGoal} from this {@linkplain AchievesManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain SpecificationGoal}.
     * @return a set of {@linkplain Role}s.
     */
    Set<Role> getAchievedBy(SpecificationGoalId id);

    /**
     * Removes an achieves relation between a {@linkplain Role} and a {@linkplain SpecificationGoal} from this {@linkplain AchievesManager}.
     *
     * @param roleId the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @param goalId the {@linkplain UniqueId} that represents the {@linkplain SpecificationGoal}.
     */
    void removeAchieves(RoleId roleId, SpecificationGoalId goalId);

    /**
     * Removes all achieves relations from this {@linkplain AchievesManager}.
     */
    void removeAllAchieves();
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/organization/AchievesManager.kt
package runtimemodels.chazm.api.organization

import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.entity.SpecificationGoal
import runtimemodels.chazm.api.id.RoleId
import runtimemodels.chazm.api.id.SpecificationGoalId
import runtimemodels.chazm.api.relation.Achieves

/**
 * The [AchievesManager] interface defines the APIs for managing a set of [Achieves] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface AchievesManager : Map<RoleId, Map<SpecificationGoalId, Achieves>> {
    /**
     * Adds an [Achieves] relation to this [AchievesManager].
     *
     * @param achieves the [Achieves] relation to add.
     */
    fun add(achieves: Achieves)

    override operator fun get(key: RoleId): Map<SpecificationGoalId, Achieves>

    operator fun get(id: SpecificationGoalId): Map<RoleId, Achieves>

    /**
     * Removes an [Achieves] relation between a [Role] and a [SpecificationGoal] from this [AchievesManager].
     *
     * @param roleId the [RoleId] that represents the [Role].
     * @param goalId the [SpecificationGoalId] that represents the [SpecificationGoal].
     * @return the [Achieves] relation that was removed.
     */
    fun remove(roleId: RoleId, goalId: SpecificationGoalId): Achieves

    /**
     * Removes all [Achieves] relations associated to a [Role] from this [AchievesManager].
     *
     * @param id the [RoleId] that represents the [Role].
     */
    fun remove(id: RoleId)

    /**
     * Removes all [Achieves] relations associated to a [SpecificationGoal] from this [AchievesManager].
     *
     * @param id the [SpecificationGoalId] that represents the [SpecificationGoal].
     */
    fun remove(id: SpecificationGoalId)
}
